File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) {
        //we need to create a queue to store all the nodes
        const queue = [{currNode: root, min: -Infinity,max: Infinity}];

        //we need to loop through all the nodes
        while(queue.length > 0){
            //we need to get the curr node out of a queue,node.min and max
            let {currNode,min,max} = queue.shift();

            //we need to comapre that node with the min and max vals
            if(currNode.val <= min || currNode.val >= max){
                return false;
            }

            //we need to push the nodes into the queue and set the new min and max
            
            //pushing the left element
            if(currNode.left != null){
                queue.push({currNode: currNode.left,min: min,max: currNode.val});
            }
            
            //pushing the right element
            if(currNode.right != null ){
                queue.push({currNode: currNode.right,min: currNode.val,max: max});
            }
        }
        return true;
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List result = new ArrayList();
        if (root == null) return result;

        Queue queue = new ArrayDeque();
        queue.offer(root);

        while(!queue.isEmpty()){
            List currentLevel = new ArrayList();
            int size = queue.size();

            for (int i = 0; i < size; i++){
                TreeNode node = (TreeNode) queue.poll();
                currentLevel.add(node.val);

                if(node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);


            }
            result.add(currentLevel);
        }
        return result;
    }
}
